mindmeld/components/question_answerer.py

```diff
@@ -15,77 +15,148 @@
-from elasticsearch import ElasticsearchException, TransportError
+import nltk
+from nltk.corpus import stopwords as nltk_stopwords
+from nltk.stem import PorterStemmer
+from nltk.tokenize import word_tokenize as nltk_word_tokenize
 
-from ..exceptions import (
-    KnowledgeBaseConnectionError,
-    KnowledgeBaseError,
-    ElasticsearchVersionError,
-)
-from ..resource_loader import ResourceLoader
 from ._config import (
-    DEFAULT_ES_QA_MAPPING,
-    DEFAULT_RANKING_CONFIG,
     get_app_namespace,
     get_classifier_config,
 )
-from ._elasticsearch_helpers import (
-    DOC_TYPE,
-    create_es_client,
-    delete_index,
-    does_index_exist,
-    get_scoped_index_name,
-    load_index,
-    create_index_mapping,
-    is_es_version_7,
-    resolve_es_config_for_version,
+from ._util import _is_module_available, _get_module_or_attr as _getattr
+from .entity_resolver import EmbedderCosSimEntityResolver, TfIdfSparseCosSimEntityResolver
+from ..core import Bunch
+from ..exceptions import (
+    ElasticsearchKnowledgeBaseConnectionError,
+    KnowledgeBaseError,
+    ElasticsearchVersionError,
 )
 from ..models import create_embedder_model
+from ..path import get_question_answerer_index_cache_file_path
+from ..resource_loader import Hasher, ResourceLoader
+
+if _is_module_available("elasticsearch"):
+    from ._elasticsearch_helpers import (
+        DOC_TYPE,
+        DEFAULT_ES_QA_MAPPING,
+        DEFAULT_ES_RANKING_CONFIG,
+        create_es_client,
+        delete_index,
+        does_index_exist,
+        get_scoped_index_name,
+        load_index,
+        create_index_mapping,
+        is_es_version_7,
+        resolve_es_config_for_version,
+    )
 
 logger = logging.getLogger(__name__)
 
-
 DEFAULT_QUERY_TYPE = "keyword"
 ALL_QUERY_TYPES = ["keyword", "text", "embedder", "embedder_keyword", "embedder_text"]
 EMBEDDING_FIELD_STRING = "_embedding"
+NON_ELASTICSEARCH_INDICES_STORAGE_PATH = os.path.join(os.path.expanduser("~"), ".cache/mindmeld")
 
 
-class QuestionAnswerer:
+class BaseQuestionAnswerer(ABC):
+
+    def __init__(self, app_path, **kwargs):
+        """
+        Args:
+            app_path (str, optional): The path to the directory containing the app's data. If
+                provided, used to obtain default `app_namespace` and QA configurations
+            app_namespace (str, optional): The namespace of the app. Used to prevent
+                collisions between the indices of this app and those of other apps.
+            config (dict, optional): The QA config if passed directly rather than loaded from the
+                app config
+            resource_loader (ResourceLoader, optional): An object which can load resources for the
+                question answerer.
+        """
+        self.app_path = app_path
+        self.app_namespace = kwargs.get("app_namespace", None)
+        self._resource_loader = kwargs.get("resource_loader", None)
+        self.__qa_config = kwargs.get("config", None)
```

Comment:
Why a double under?

Reply:
Just to ensure it is not used in derived classes. One must use `_query_type` or `_query_settings` properties of base class instead. Helps to keep things tractable I feel. Also, using `_query_type` is safer as it avoids unnecessary errors due to unknown query_type strings.